File: OpenGSGLibrary/GameDataManager/Province.vb
'    OpenSGSGLibrary is an open-source library for Grand Strategy Games
'    Copyright (C) 2019  Torben Kneesch
'
'    This program is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    This program is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Namespace WorldData

    ''' <summary>
    ''' Base class for provinces. Handles the most basic province properties.
    ''' </summary>
    Public Class Province
        Inherits GameObject

        Public Function GetId() As Integer
            Return id_
        End Function

        Public Function GetName() As String
            Return name_
        End Function

        Public Function GetController() As String
            Return controller_
        End Function

        Public Function GetOwner() As String
            Return owner_
        End Function

        ''' <summary>
        ''' Sets the properties of a province from the parser data of the province file.
        ''' This method handles province id, name, controller's country tag and owner's country tag.
        ''' Should be inherited to handle more game-specific properties.
        ''' </summary>
        ''' <param name="fileName"></param>
        ''' <param name="parsedData"></param>
        Public Overrides Sub SetData(fileName As String, parsedData As Dictionary(Of String, Object))
            MyBase.SetData(fileName, parsedData)
            Dim fileNameParts As String() = GameObjectFactory.ExtractFromFilename(fileName)
            id_ = Val(fileNameParts(0))
            name_ = fileNameParts(1)

            controller_ = parsedData("controller")
            owner_ = parsedData("owner")
        End Sub

        Private id_ As Integer
        Private name_ As String
        Private controller_ As String
        Private owner_ As String

    End Class

End Namespace

File: OpenGSGLibrary/GameDataManager/Province.cs
namespace OpenGSGLibrary.GameDataManager
{
    /// <summary>
    /// Base class for provinces. Keeps model as a simple POCO — notifications live on ViewModels.
    /// </summary>
    public class Province : GameObject
    {
        public int Id { get; private set; }
        public string Name { get; private set; } = string.Empty;
        public string Controller { get; set; } = string.Empty;
        public string Owner { get; set; } = string.Empty;

        /// <summary>
        /// List of country tags that consider this province their core territory.
        /// Used for territorial claims, nationalism, and liberation mechanics.
        /// </summary>
        public List<string> Cores { get; set; } = new List<string>();

        public override void SetData(string fileName, ILookup<string, object> parsedData)
        {
            base.SetData(fileName, parsedData);
            var fileNameParts = GameObjectFactory.ExtractFromFilename(fileName);
            Id = Convert.ToInt32(fileNameParts[0]);
            Name = fileNameParts.Length > 1 ? fileNameParts[1] : string.Empty;

            Controller = parsedData.Contains("controller")
                ? parsedData["controller"].Single()?.ToString() ?? string.Empty
                : string.Empty;
            Owner = parsedData.Contains("owner")
                ? parsedData["owner"].Single()?.ToString() ?? string.Empty
                : string.Empty;

            // Parse cores (using add_core keyword)
            if (parsedData.Contains("add_core"))
            {
                foreach (var coreTag in parsedData["add_core"])
                {
                    var tag = coreTag?.ToString();
                    if (!string.IsNullOrEmpty(tag) && !Cores.Contains(tag))
                    {
                        Cores.Add(tag);
                    }
                }
            }
        }
    }
}
